### PR TITLE
Issue #9 - Configuration Redesign

### PR DESCRIPTION
**Pull request by reynoldtan**
- Set up form as per mockup
- Populated fields
- Save value entered

### CODE REVIEW

File: include/analyzedphenotypes.admin.form.inc [hunk 1]
```diff
@@ -1062,409 +1062,274 @@ function analyzedphenotypes_dragdrop($form, &$form_state, $source_name) {
...
+    '<u>Crop Ontology</u>:' .
+      '<br />Our experience with breeders has led us to recommend using the trait names your breeder(s) already use in the Trait Vocabulary and then linking them to a more generic crop ontology such as those provided by <u><em>' . l('cropontology.org', 'http://cropontology.org') . '</em></u> to facilitate sharing. If you decide to go this route, you can set the species specific crop ontology here and on upload suitable terms will be suggested based on pattern matching.',
+  );
 
-      // CV and Genus, related (ontology).
-      $var_cv = $vars['ap_cv'];
-      $var_cv_value = variable_get($var_cv);
+  $form[$frameset_to]['config_instructions'] = array(
+    '#markup' => theme('item_list', array('items' => $items)),
+  );
```
Comment (laceysanderson) [inline include/analyzedphenotypes.admin.form.inc]: It's better to use:
```php
  $form[$frameset_to]['config_instructions'] = array(
      '#theme' => 'item_list',
      '#items' => $items,
  );
```

File: include/analyzedphenotypes.admin.form.inc [hunk 2]
```diff
@@ -1062,409 +1062,274 @@ function analyzedphenotypes_dragdrop($form, &$form_state, $source_name) {
...
         );
+
+        $field_name = $basename . $genus_key . '_' . $var_token[$j][0];
       }
 
-      // Allow new traits?
-      $var_allow_new = $vars['ap_allow_new'];
-      $var_allow_new_value = variable_get($var_allow_new);
-      $default_allow_new = ($var_allow_new_value == 'not set') ? 0 : $var_allow_new_value;
+      $fields[ 'field_row_' . $i ][$field_name] = $field_prop;
```
Comment (laceysanderson) [inline include/analyzedphenotypes.admin.form.inc]: Inconsistency: on line 1124 you set the row to be `'ap_field_row_' . $i` but now here you assign the row to `'field_row_' . $i`

File: analyzedphenotypes.module [hunk 3]
```diff
@@ -167,10 +135,39 @@ function analyzedphenotypes_theme($existing, $type, $theme, $path) {
...
+  // Form table.
+  $items['analyzedphenotypes_form_table'] = array(
+    'render element' => 'form',
+  );
+
   return $items;
 }
 
 
+function theme_analyzedphenotypes_form_table(&$variables) {
```
Comment (laceysanderson) [inline analyzedphenotypes.module]: Where is the doc block?

File: analyzedphenotypes.module [hunk 4]
```diff
@@ -311,6 +308,20 @@ function analyzedphenotypes_data_json($page_argument_5, $page_argument_6, $key =
 
       break;
 
+    //
```
Comment (laceysanderson) [inline analyzedphenotypes.module:388]: Make sure that each case in this function has documentation of:
- where it is used
- what it is for
- what you expect the options to look like

Additionally, the design of this function makes me nervous. In my experience it is too easy to forget a `break;` and end up with additional options which might be hard to debug. Perhaps, at a minimum you should add a `$data_JSON = array();` at the beginning of each case to ensure there is no flow over if a break is missed. Personally, I would have made each function pulling options (e.g. analyzedphenotypes_cvprop) simply output the JSON as well.

Comment (laceysanderson) [reply analyzedphenotypes.module:388]: Upon further digging, I noticed this design of using switch statements to combine multiple, loosely related pieces of functionality into a single function is prevalent. Thus for this PR, keep the design with the switch and simply add the `$data_JSON = array();` at the beginning of each case.

**Review by laceysanderson — CHANGES_REQUESTED** [review]

File: analyzedphenotypes.module (hunk 4 above)
Comment (reynoldtan) [reply analyzedphenotypes.module:388]: Updated doc block and removed unused case.

**Review by laceysanderson — APPROVED** [review]
Code Changes look good :-)

**Comment by laceysanderson** [comment]
New configuration form looks good 👍 

However, once I configure the Trait Ontologies for Lentil and then try the loader, I'm still told that `System settings not set. Please contact the administrator of this site.`. Administrators likely won't want to enable phenotypes for all organisms in their database. Therefore, this form needs to all some organism to not have any trait information set. However, if either the trait cv or db is set for a given organism then both need to be. Furthermore, the crop ontology should be optional in all cases.

**Comment by laceysanderson** [comment]
The validation on the upload page will addressed as part of separate issue/PR #10.

One last thing for this PR: In-form validation (using `form_set_error()` ) is required for
 - Trait ontologies: provide error if only one of trait cv + db is supplied for a given organism
 - CV Terms: term existence should be checked on submit. Currently I can submit `sarah` with no repercussions despite the fact that it doesn't exist and no cv is indicated.

**Comment by reynoldtan** [comment]
Please note that this PR requires PR - Config. Page Add Default Controlled Vocabulary Terms #13.

File: include/analyzedphenotypes.admin.form.inc [hunk 5]
```diff
@@ -1061,410 +1061,348 @@ function analyzedphenotypes_dragdrop($form, &$form_state, $source_name) {
...
-        );
+  // Test if value entered (field for terms are autocomplete).
+  foreach($terms as $term => $t) {
+    $field_term = $form_state['values'][$t];
 
-        return $form;
-  }
+    if (empty($field_term)) {
+      // Field is empty. Mark field.
+      form_set_error($t, '');
```
Comment (laceysanderson) [inline include/analyzedphenotypes.admin.form.inc]: The message should always be directly in the form_set_error and as specific as possible. Especially on an administration page ;-) For example, a good message for this one would be "You must set the controlled vocabulary term for [category; e.g. related] as it is required for the proper functioning of this module."

File: include/analyzedphenotypes.admin.form.inc [hunk 6]
```diff
@@ -1061,410 +1061,348 @@ function analyzedphenotypes_dragdrop($form, &$form_state, $source_name) {
...
+    }
+    else {
+      // Check if term in cvterm table.
+      $is_term = analyzedphenotypes_cvprop('term_cv', array(
+        'term_cv' => $field_term,
+      ));
 
+      if (is_null($is_term)) {
+        // Invalid term. Mark field.
+        form_set_error($t, '');
```
Comment (laceysanderson) [inline include/analyzedphenotypes.admin.form.inc]: See last comment. A good message would be "[term entered] is not a valid controlled vocabulary term. Please confirm that it already exists keeping in mind that case and whitespace are important."

**Review by laceysanderson — CHANGES_REQUESTED** [review]
I'm also receiving errors on submit of the controlled vocabularies due to them being set to the cvterm_id. This makes it difficult to test that validation but we'll move forward anyway and I'll test as part of the next PR.

For This PR, all that's required for merging is fixing of the form_set_error ()messages to those suggested below :-)

**Comment by reynoldtan** [comment]
Made error messages specific.
Sample screenshot of validation messages.
vr

**Comment by laceysanderson** [comment]
Perfect 👍

**Review by laceysanderson — APPROVED** [review]

**Pull request merged**